crypto/src/batch_contribution.rs

```diff
@@ -17,17 +19,28 @@ impl BatchContribution {
     }
 
     #[instrument(level = "info", skip_all, fields(n=self.contributions.len()))]
-    pub fn add_entropy<E: Engine>(&mut self, entropy: [u8; 32]) -> Result<(), CeremoniesError> {
-        let entropies = derive_entropy(entropy, self.contributions.len());
-        self.contributions
+    pub fn add_entropy<E: Engine>(&mut self, mut entropy: [u8; 32]) -> Result<(), CeremoniesError> {
+        // make sure that the entropy passed in is not prior zeroized
+        assert_ne!(entropy, [0; 32]);
+
+        let mut entropies = derive_entropy(entropy, self.contributions.len());
+        let res = self
+            .contributions
             .par_iter_mut()
-            .zip(entropies)
+            .zip(&mut entropies)
             .enumerate()
             .try_for_each(|(i, (contribution, entropy))| {
-                contribution
+                let res = contribution
                     .add_entropy::<E>(entropy)
-                    .map_err(|e| CeremoniesError::InvalidCeremony(i, e))
-            })
+                    .map_err(|e| CeremoniesError::InvalidCeremony(i, e));
+                // make sure the entropy has been zeroized
+                assert_eq!(*entropy, [0; 32]);
+                res
+            });
+
+        // zeroize the toxic waste
+        entropy.zeroize();
```

Comment:
Using a `type Entropy = Zeroizing<[u8; 32]>` will make sure we can not forget this statement (for example if we early exit due to an error)